core/ffmpeg/hlsVariantWriteMonitor.go

```diff
@@ -0,0 +1,95 @@
+package ffmpeg
+
+import (
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+type hlsVariantWriteMonitor struct {
+	playlistEvents chan hlsWriteEvent
+	segmentEvents  chan hlsWriteEvent
+	logger         *log.Logger
+}
+
+type hlsWriteEvent struct {
+	Dir  string
+	Time time.Time
+}
+
+func newHlsVariantWriteMonitor(ctx context.Context, logger *log.Logger, notificationThreshold time.Duration) *hlsVariantWriteMonitor {
+	h := &hlsVariantWriteMonitor{make(chan hlsWriteEvent), make(chan hlsWriteEvent), logger}
+	go h.run(ctx, notificationThreshold)
+	return h
+}
+
+func (h *hlsVariantWriteMonitor) SegmentWritten(dir string, when time.Time) {
+	select {
+	case h.segmentEvents <- hlsWriteEvent{dir, when}:
+	case <-time.After(50 * time.Millisecond):
+		h.logger.Errorf("unable to monitor segment write %q / %v", dir, when)
+	}
+}
+
+func (h *hlsVariantWriteMonitor) VariantPlaylistWritten(dir string, when time.Time) {
+	select {
+	case h.playlistEvents <- hlsWriteEvent{dir, when}:
+	case <-time.After(50 * time.Millisecond):
+		h.logger.Errorf("unable to monitor playlist write %q / %v", dir, when)
+	}
+}
+
+type hlsVariantWriteHistory struct {
+	lastSegmentWrite  time.Time
+	lastPlaylistWrite time.Time
+}
+
+// Run watches the handler traffic for lagging playlist writes
+func (h *hlsVariantWriteMonitor) run(ctx context.Context, notificationThreshold time.Duration) {
+	var (
+		event  hlsWriteEvent
+		drift  time.Duration
+		viable bool
+
+		variant  *hlsVariantWriteHistory
+		ok       bool
+		variants map[string]*hlsVariantWriteHistory = make(map[string]*hlsVariantWriteHistory)
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case event = <-h.playlistEvents:
+			variant, ok = variants[event.Dir]
+			if !ok {
+				variant = &hlsVariantWriteHistory{event.Time, event.Time}
+			} else {
+				variant.lastPlaylistWrite = event.Time
+			}
+			variants[event.Dir] = variant
+			drift = 0
+			viable = false
+			continue
+
+		case event = <-h.segmentEvents:
+			variant, ok = variants[event.Dir]
+			if !ok {
+				variant = &hlsVariantWriteHistory{event.Time, event.Time}
+				drift = 0
+			} else {
+				drift = event.Time.Sub(variant.lastPlaylistWrite)
+				viable = event.Time.Sub(variant.lastSegmentWrite) < notificationThreshold
```

Comment:
If the time between segments has been delayed long enough, I don't count that as a playlist write miss, since it could be from a stream disconnect and then reconnect.

Reply:
I'm probably missing something, but it seems like `viable` is always ending up as `false` as it's using the same `notificationThreshold` that it's using for notifying against.  Sorry if I missed the whole point of something :)

Reply:
🤔  Not positive I understand.  `notificationThreshold` is constant in this scope (it's assuming all variants write HLS segments at the same rate). So I'm taking the current segment write time - last segment write and seeing if it's greater than the allowed drift of playlist/segment writes.  

So a segment written > notificationThreshold from playlist triggers the log, but if no segments have been written in that same window, I assume that is due to a disconnect in streaming.  `Test_hlsVariantWriteMonitor_segmentDelay` covers this bit.

I think the reset on disconnect is probably the better option, I'll see if I can work that in.